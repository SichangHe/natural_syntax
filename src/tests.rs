use super::*;

use std::time::Instant;

use insta::assert_debug_snapshot;
use tracing::{info, Level};
use tracing_subscriber::EnvFilter;

#[test]
fn paragraph_pos() {
    init_tracing();
    let start = Instant::now();
    let model = POSModel::try_default().unwrap();
    info!("Took {}ms to load the model.", start.elapsed().as_millis());
    let input = "Extracts Part of Speech tags (Noun, Verb, Adjectiveâ€¦) from text. A lightweight pretrained model using MobileBERT is available for English.";
    let start = Instant::now();
    let output = model.predict(input);
    info!("Took {}ms to predict.", start.elapsed().as_millis());
    let mut parsed = output.collect::<Result<Vec<_>, _>>().unwrap();
    round_scores(&mut parsed);
    assert_debug_snapshot!(parsed);
}

#[test]
fn markdown_pos() {
    init_tracing();
    let start = Instant::now();
    let model = POSModel::try_default().unwrap();
    info!("Took {}ms to load the model.", start.elapsed().as_millis());
    let input = "The easiest way of getting started is the [rustler Elixir library](https://hex.pm/packages/rustler).

- Add the [rustler Elixir library](https://hex.pm/packages/rustler) as a
  dependency of your project.
- Run `mix rustler.new` to generate a new NIF in your project. Follow the
  instructions.
- If you are already using [`serde`](https://serde.rs) and/or have been using
  `serde_rustler` before, please enable the `serde` feature in your NIF crate's
  `Cargo.toml` on the `rustler` dependency.
";
    let start = Instant::now();
    let output = model.predict(input);
    info!("Took {}ms to predict.", start.elapsed().as_millis());
    let mut parsed = output.collect::<Result<Vec<_>, _>>().unwrap();
    round_scores(&mut parsed);
    assert_debug_snapshot!(parsed);
}

const PRECISION: f64 = 1e-3;

fn round_scores(predictions: &mut [POSToken]) {
    predictions
        .iter_mut()
        .for_each(|token| token.score = (token.score / PRECISION).round() * PRECISION)
}

fn init_tracing() {
    _ = tracing_subscriber::fmt()
        .with_env_filter(
            EnvFilter::builder()
                .with_default_directive(Level::INFO.into())
                .from_env_lossy(),
        )
        .try_init();
}
