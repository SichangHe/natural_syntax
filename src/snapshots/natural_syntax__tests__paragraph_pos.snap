---
source: src/tests.rs
expression: parsed
---
[
    POSToken {
        word: "Extracts",
        score: 0.676,
        tag: VBZ,
        label_index: 30,
        index: 1,
        word_index: 1,
        offset_begin: 0,
        offset_end: 8,
    },
    POSToken {
        word: "Part",
        score: 0.981,
        tag: NN,
        label_index: 11,
        index: 2,
        word_index: 2,
        offset_begin: 9,
        offset_end: 13,
    },
    POSToken {
        word: "of",
        score: 0.999,
        tag: IN,
        label_index: 6,
        index: 3,
        word_index: 3,
        offset_begin: 14,
        offset_end: 16,
    },
    POSToken {
        word: "Speech",
        score: 0.902,
        tag: NN,
        label_index: 11,
        index: 4,
        word_index: 4,
        offset_begin: 17,
        offset_end: 23,
    },
    POSToken {
        word: "tags",
        score: 0.925,
        tag: NNS,
        label_index: 14,
        index: 5,
        word_index: 5,
        offset_begin: 24,
        offset_end: 28,
    },
    POSToken {
        word: "(",
        score: 0.668,
        tag: IN,
        label_index: 6,
        index: 6,
        word_index: 6,
        offset_begin: 29,
        offset_end: 30,
    },
    POSToken {
        word: "Noun",
        score: 0.712,
        tag: NN,
        label_index: 11,
        index: 7,
        word_index: 7,
        offset_begin: 30,
        offset_end: 34,
    },
    POSToken {
        word: ",",
        score: 0.363,
        tag: CC,
        label_index: 1,
        index: 8,
        word_index: 8,
        offset_begin: 34,
        offset_end: 35,
    },
    POSToken {
        word: "Verb",
        score: 0.798,
        tag: NN,
        label_index: 11,
        index: 9,
        word_index: 9,
        offset_begin: 36,
        offset_end: 40,
    },
    POSToken {
        word: ",",
        score: 0.23900000000000002,
        tag: CC,
        label_index: 1,
        index: 10,
        word_index: 10,
        offset_begin: 40,
        offset_end: 41,
    },
    POSToken {
        word: "Adjective",
        score: 0.467,
        tag: JJ,
        label_index: 7,
        index: 11,
        word_index: 11,
        offset_begin: 42,
        offset_end: 51,
    },
    POSToken {
        word: "â€¦",
        score: 0.311,
        tag: CD,
        label_index: 2,
        index: 12,
        word_index: 12,
        offset_begin: 51,
        offset_end: 52,
    },
    POSToken {
        word: ")",
        score: 0.315,
        tag: DT,
        label_index: 3,
        index: 13,
        word_index: 13,
        offset_begin: 52,
        offset_end: 53,
    },
    POSToken {
        word: "from",
        score: 0.994,
        tag: IN,
        label_index: 6,
        index: 14,
        word_index: 14,
        offset_begin: 54,
        offset_end: 58,
    },
    POSToken {
        word: "text",
        score: 0.763,
        tag: NN,
        label_index: 11,
        index: 15,
        word_index: 15,
        offset_begin: 59,
        offset_end: 63,
    },
    POSToken {
        word: ".",
        score: 0.5760000000000001,
        tag: WDT,
        label_index: 31,
        index: 16,
        word_index: 16,
        offset_begin: 63,
        offset_end: 64,
    },
    POSToken {
        word: "A",
        score: 0.921,
        tag: DT,
        label_index: 3,
        index: 17,
        word_index: 17,
        offset_begin: 65,
        offset_end: 66,
    },
    POSToken {
        word: "lightweight",
        score: 0.96,
        tag: JJ,
        label_index: 7,
        index: 18,
        word_index: 18,
        offset_begin: 67,
        offset_end: 78,
    },
    POSToken {
        word: "pretrained",
        score: 0.088,
        tag: JJ,
        label_index: 7,
        index: 19,
        word_index: 19,
        offset_begin: 79,
        offset_end: 89,
    },
    POSToken {
        word: "model",
        score: 0.987,
        tag: NN,
        label_index: 11,
        index: 22,
        word_index: 20,
        offset_begin: 90,
        offset_end: 95,
    },
    POSToken {
        word: "using",
        score: 0.916,
        tag: VBG,
        label_index: 27,
        index: 23,
        word_index: 21,
        offset_begin: 96,
        offset_end: 101,
    },
    POSToken {
        word: "MobileBERT",
        score: 0.558,
        tag: NNP,
        label_index: 12,
        index: 24,
        word_index: 22,
        offset_begin: 102,
        offset_end: 112,
    },
    POSToken {
        word: "is",
        score: 0.996,
        tag: VBZ,
        label_index: 30,
        index: 26,
        word_index: 23,
        offset_begin: 113,
        offset_end: 115,
    },
    POSToken {
        word: "available",
        score: 0.668,
        tag: JJ,
        label_index: 7,
        index: 27,
        word_index: 24,
        offset_begin: 116,
        offset_end: 125,
    },
    POSToken {
        word: "for",
        score: 0.999,
        tag: IN,
        label_index: 6,
        index: 28,
        word_index: 25,
        offset_begin: 126,
        offset_end: 129,
    },
    POSToken {
        word: "English",
        score: 0.88,
        tag: NNP,
        label_index: 12,
        index: 29,
        word_index: 26,
        offset_begin: 130,
        offset_end: 137,
    },
    POSToken {
        word: ".",
        score: NaN,
        tag: CC,
        label_index: 1,
        index: 30,
        word_index: 27,
        offset_begin: 137,
        offset_end: 138,
    },
]
